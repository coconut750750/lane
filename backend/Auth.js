import firebase from 'firebase';

import { clientId } from 'lane/config';

export function getUserID() {
    return firebase.auth().currentUser.uid;
}

function isUserEqual(googleUser, firebaseUser) {
    if (firebaseUser) {
        var providerData = firebaseUser.providerData;
        for (var i = 0; i < providerData.length; i++) {
            if (providerData[i].providerId === firebase.auth.GoogleAuthProvider.PROVIDER_ID &&
                providerData[i].uid === googleUser.getBasicProfile().getId()) {
                return true;
            }
        }
    }
    return false;
}

function onSignIn(googleUser) {
    // We need to register an Observer on Firebase Auth to make sure auth is initialized.
    var unsubscribe = firebase.auth().onAuthStateChanged(firebaseUser => {
        unsubscribe();
        // Check if we are already signed-in Firebase with the correct user.
        if (!isUserEqual(googleUser, firebaseUser)) {
            // Build Firebase credential with the Google ID token.
            var credential = firebase.auth.GoogleAuthProvider.credential(
                googleUser.idToken,
                googleUser.accessToken
             );
            // Sign in with credential from the Google user.
            firebase
                .auth()
                .signInWithCredential(credential)
                .then(function(result) {
                    if (result.additionalUserInfo.isNewUser) {
                        firebase.database()
                            .ref('users')
                            .child(result.user.uid)
                            .set({
                                email: result.user.email,
                                profile: result.additionalUserInfo.profile.picture,
                                firstname: result.additionalUserInfo.profile.given_name,
                                lastname: result.additionalUserInfo.profile.family_name
                            });
                        firebase.database()
                            .ref('emails')
                            .child(result.user.uid)
                            .set(result.user.email);
                    }
                })
                .catch(function(error) {
                    // Handle Errors here.
                    var errorCode = error.code;
                    var errorMessage = error.message;
                    // The email of the user's account used.
                    var email = error.email;
                    // The firebase.auth.AuthCredential type that was used.
                    var credential = error.credential;
                });
        } else {
            console.log('User already signed-in Firebase.');
        }
    });
}

export async function signInWithGoogleAsync() {
    try {
        const result = await Expo.Google.logInAsync({
            behavior: 'web',
            androidClientId: clientId.android,
            iosClientId: clientId.ios,
            scopes: ['profile', 'email']
        });

        if (result.type === 'success') {
            onSignIn(result);
            return result.accessToken;
        } else {
            return {canceled: true};
        }
    } catch (e) {
        return {error: true};
    }
}

export function signOut() {
    firebase.auth().signOut();
}

export function checkIfLoggedIn(checkComplete) {
    firebase.auth().onAuthStateChanged(user => {
        checkComplete(user != undefined);
    });
}

export async function getIdFromEmail(email, onComplete) {
    firebase.database().ref('emails').once('value', snapshot => {
        var emails = snapshot.val();
        for (var id in emails) {
            if (emails[id] === email) {
                onComplete(id);
                return;
            }
        }
        onComplete(undefined);
    });
}