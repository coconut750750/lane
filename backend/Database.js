import firebase from 'firebase';
import { convertTimestampToDateString } from 'lane/utils/utils';

let USERS = 'users';
let LANES = 'lanes';
let SHARED = 'shared';

export async function addLaneToUser(userId, laneId) {
    await firebase.database()
        .ref(USERS).child(userId)
        .child(LANES).child(laneId)
        .set(0);
}

export async function removeLaneFromUser(userId, laneId) {
    await firebase.database()
        .ref(USERS).child(userId)
        .child(LANES).child(laneId)
        .remove();
}

export async function removeUser(userId) {
    await firebase.database()
        .ref(USERS).child(userId)
        .remove();
}

export async function addUserToLane(userId, laneId) {
    await firebase.database()
        .ref(LANES).child(laneId)
        .child(SHARED).child(userId)
        .set(0);
}

export async function removeUserFromLane(userId, laneId) {
    await firebase.database()
        .ref(LANES).child(laneId)
        .child(SHARED).child(userId)
        .remove();
}

export async function removeAllSharedUsers(laneObj) {
    if (laneObj.shared != null) {
        await Promise.all(
            Object.keys(laneObj.shared).map( id => 
                removeLaneFromUser(id, laneObj.id)
            )
        );
    }
}

export async function removeLane(laneId) {
    await firebase.database()
        .ref(LANES).child(laneId)
        .remove();
}

export async function addPhoto(blob, photo, laneId) {
    let photoId = photo.md5;

    const snapshot = await firebase.storage()
        .ref()
        .child(laneId)
        .child(photoId)
        .put(blob);
    blob.close();

    const photoUrl = await snapshot.ref.getDownloadURL();

    await firebase.database()
        .ref(LANES).child(laneId)
        .child('photos').child(photoId)
        .set({
            uri: photoUrl,
            width: photo.image.width,
            height: photo.image.height,
            timestamp: convertTimestampToDateString(photo.timestamp),
        });
}

export async function removePhoto(photoId, laneId) {
    await firebase.storage()
        .ref()
        .child(laneId)
        .child(photoId)
        .delete();

    await firebase.database()
        .ref(LANES).child(laneId)
        .child('photos').child(photoId)
        .remove();
}

export async function removeAllLanePhotos(laneObj) {
    await Promise.all(
        Object.keys(laneObj.photos).map( photoId => 
            firebase.storage().ref().child(laneObj.id).child(photoId).delete()
        )
    );
}

export async function pushLane(ownerId, laneObj) {
    laneObj.owner = ownerId;
    return await firebase.database().ref(LANES).push(laneObj).key;
}

export async function onLaneUpdate(processLanes) {
    var userid = firebase.auth().currentUser.uid;
    firebase.database().ref(USERS).child(userid).child(LANES).on('value', datasnapshot => {
        var lanes = datasnapshot.val();
        var laneObjs = {};

        if (!lanes) {
            processLanes(laneObjs);
        } else {
            var markings = Object.keys(lanes).map( laneId =>
                firebase.database().ref(LANES).child(laneId).once('value', lanesnapshot => {
                    var lane = lanesnapshot.val();
                    laneObjs[laneId] = {...lane, id: laneId};
                })
            );
            
            Promise.all(markings).then(lanes => {
                processLanes(laneObjs);
            });
        }
    });
}

export async function uploadImageAsync(laneId, photo) {
    const blob = await new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.onload = function() {
            resolve(xhr.response);
        };
        xhr.onerror = function(e) {
            console.log(e);
            reject(new TypeError('Network request failed'));
        };
        xhr.responseType = 'blob';
        xhr.open('GET', photo.image.uri, true);
        xhr.send(null);
    });

    await addPhoto(blob, photo, laneId);
}

export async function shareLane(laneId, otherUserId) {
    await Promise.all([
        addLaneToUser(otherUserId, laneId),
        addUserToLane(otherUserId, laneId)
    ]);
}

export function deleteLane(laneId, onComplete) {
    firebase.database().ref(LANES).child(laneId).once('value', lanesnapshot => {
        var laneObj = lanesnapshot.val();
        laneObj.id = laneId;
        Promise.all([
            removeAllSharedUsers(laneObj),
            removeLaneFromUser(laneObj.owner, laneId),
            removeAllLanePhotos(laneObj),
            removeLane(laneId)
        ]).then(_ => {
            onComplete();
        });
    });
}