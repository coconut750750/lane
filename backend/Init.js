import firebase from 'firebase';

import { firebaseConfig } from '../config'

export default function init() {
    firebase.initializeApp(firebaseConfig)
}
