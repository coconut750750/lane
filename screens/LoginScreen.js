import React, { Component } from 'react';
import { View, Text, Button, ActivityIndicator, StyleSheet } from 'react-native';

import { signInWithGoogleAsync } from 'lane/backend/Auth';

import Colors from 'lane/constants/Colors';

export default class LoginScreen extends Component {
    constructor(props) {
        super(props);
        this.state = {
            didAttemptLogin: false
        };
    }

    attemptSignIn() {
        signInWithGoogleAsync();
        this.setState({
            didAttemptLogin: true
        });
    }
    
    renderLogin() {
        return (
            <View style={styles.container}>
                <Button
                    title="Sign in with Google"
                    onPress={ () => this.attemptSignIn() }
                />
            </View>
        );
    }

    renderLoading() {
        return (
            <View style={styles.container}>
                <ActivityIndicator 
                    size="large"
                    color={Colors.primary}
                />
            </View>
        );
    }

    render() {
        if (!this.state.didAttemptLogin) {
            return this.renderLogin();
        } else {
            return this.renderLoading();
        }
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center'
    }
})