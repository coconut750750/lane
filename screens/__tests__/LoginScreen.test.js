import React from 'react'
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import * as auth from 'lane/backend/Auth';

import LoginScreen from '../LoginScreen'

jest.mock("lane/backend/Auth");

Enzyme.configure({ adapter: new Adapter() });

describe('LoginScreen', () => {
    it('renders login screen', () => {
        const component = shallow(<LoginScreen />);
        const instance = component.instance();
        instance.state.didAttemptLogin = false;
        const screen = renderer.create(instance.render()).toJSON();
        expect(screen).toMatchSnapshot();
    });

    it('renders loading screen', () => {
        const component = shallow(<LoginScreen />);
        const instance = component.instance();
        instance.state.didAttemptLogin = true;
        const screen = renderer.create(instance.render()).toJSON();
        expect(screen).toMatchSnapshot();
    });

    it('attempts to login', () => {
        const component = shallow(<LoginScreen />);
        const instance = component.instance();
        expect(instance.state.didAttemptLogin).toBeFalsy();
        instance.attemptSignIn();

        expect(auth.signInWithGoogleAsync).toHaveBeenCalled();
        expect(instance.state.didAttemptLogin).toBeTruthy();
    });
});