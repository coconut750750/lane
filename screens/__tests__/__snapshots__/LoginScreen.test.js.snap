// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders loading screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#6200ee"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`LoginScreen renders login screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Sign in with Google
      </Text>
    </View>
  </View>
</View>
`;
